src/util/newconfig/ConfigValue.hpp

```diff
@@ -83,47 +61,101 @@ class ConfigValue {
     {
-        setValue(value);
+        auto const err = checkTypeConsistency(type_, value);
+        ASSERT(!err.has_value(), "{}", err->error);
+        value_ = value;
         return *this;
     }
 
     /**
-     * @brief Gets the config type
+     * @brief Sets the value current ConfigValue given by the User's defined value
      *
-     * @return The config type
+     * @param value The value to set
+     * @param key The Config key associated with the value. Optional to include; Used for debugging message to user.
+     * @return optional Error if user tries to set a value of wrong type or not within a constraint
      */
-    [[nodiscard]] constexpr ConfigType
-    type() const
+    [[nodiscard]] std::optional<Error>
+    setValue(Value value, std::optional<std::string_view> key = std::nullopt)
     {
-        return type_;
+        auto err = checkTypeConsistency(type_, value);
+        if (err.has_value()) {
+            if (key.has_value())
+                err->error = fmt::format("{} {}", key.value(), err->error);
+            return err;
+        }
+
+        if (cons_.has_value()) {
+            auto constraintCheck = cons_->get().checkConstraint(value);
+            if (constraintCheck.has_value()) {
+                if (key.has_value())
+                    constraintCheck->error = fmt::format("{} {}", key.value(), constraintCheck->error);
+                return constraintCheck;
+            }
+        }
+        value_ = value;
+        return std::nullopt;
     }
 
     /**
-     * @brief Sets the minimum value for the config
+     * @brief Assigns a constraint to the ConfigValue.
      *
-     * @param min The minimum value
-     * @return Reference to this ConfigValue
+     * This method associates a specific constraint with the ConfigValue.
+     * If the ConfigValue already holds a value, the method will check whether
+     * the value satisfies the given constraint. If the constraint is not satisfied,
+     * an assertion failure will occur with a detailed error message.
+     *
+     * @param cons The constraint to be applied to the ConfigValue.
+     * @return A reference to the modified ConfigValue object.
      */
     [[nodiscard]] constexpr ConfigValue&
-    min(std::uint32_t min)
+    withConstraint(Constraint const& cons)
     {
-        min_ = min;
+        cons_ = std::reference_wrapper<Constraint const>(cons);
+        ASSERT(cons_.has_value(), "Constraint must be defined");
+
+        if (value_.has_value()) {
+            auto const& temp = cons_.value().get();
+            auto const& result = temp.checkConstraint(value_.value());
+            if (result.has_value()) {
+                // useful for specifying clear Error message
+                std::string type;
+                std::visit(
+                    util::OverloadSet{
+                        [&type](bool tmp) { type = fmt::format("bool {}", tmp); },
+                        [&type](std::string tmp) { type = fmt::format("string {}", tmp); },
```

Comment:
For more complex types like `std::string` it's probably better to avoid a copy here by using `std::string const& tmp`.